src/rgw/rgw_op.cc

```diff
@@ -3564,63 +3564,68 @@ void RGWCreateBucket::execute(optional_yield y)
-      ldpp_dout(this, 0) << "could not find zonegroup "
-          << createparams.zonegroup_id << " in current period" << dendl;
-      op_ret = -ENOENT;
-      return;
-    }
-    bucket_zonegroup = &z->second;
-  } else if (createparams.zonegroup_id != my_zonegroup.id) {
-    ldpp_dout(this, 0) << "zonegroup does not match current zonegroup "
-        << createparams.zonegroup_id << dendl;
-    op_ret = -ENOENT;
-    return;
-  }
 
-  // validate the LocationConstraint
+  // Validate LocationConstraint if it's provided and enforcement is strict
   if (!location_constraint.empty() && !relaxed_region_enforcement) {
-    // on the master zonegroup, allow any valid api_name. otherwise it has to
-    // match the bucket's zonegroup
-    if (period && my_zonegroup.is_master) {
-      if (!period->period_map.zonegroups_by_api.count(location_constraint)) {
+    if (period) {
+      auto location_iter = period->period_map.zonegroups_by_api.find(location_constraint);
+      if (location_iter == period->period_map.zonegroups_by_api.end()) {
         ldpp_dout(this, 0) << "location constraint (" << location_constraint
             << ") can't be found." << dendl;
         op_ret = -ERR_INVALID_LOCATION_CONSTRAINT;
         s->err.message = "The specified location-constraint is not valid";
         return;
       }
-    } else if (bucket_zonegroup->api_name != location_constraint) {
+      bucket_zonegroup = &location_iter->second;
+    } else if (location_constraint != my_zonegroup.api_name) {
       ldpp_dout(this, 0) << "location constraint (" << location_constraint
-          << ") doesn't match zonegroup (" << bucket_zonegroup->api_name
-          << ')' << dendl;
-      op_ret = -ERR_INVALID_LOCATION_CONSTRAINT;
-      s->err.message = "The specified location-constraint is not valid";
+          << ") doesn't match zonegroup (" << my_zonegroup.api_name << ")" << dendl;
+      op_ret = -ERR_ILLEGAL_LOCATION_CONSTRAINT_EXCEPTION;
+      s->err.message = fmt::format("The {} location constraint is incompatible "
+                                   "for the region specific endpoint this request was sent to.", 
+                                   location_constraint);
       return;
     }
+  } 
+  // If it's a system request, use the provided zonegroup if available
+  else if (s->system_request && !rgwx_zonegroup.empty()) {
+    if (period) {
+      auto zonegroup_iter = period->period_map.zonegroups.find(rgwx_zonegroup);
+      if (zonegroup_iter == period->period_map.zonegroups.end()) {
+        ldpp_dout(this, 0) << "could not find zonegroup " << rgwx_zonegroup 
+            << " in current period" << dendl;
+        op_ret = -ENOENT;
+        return;
+      }
+      bucket_zonegroup = &zonegroup_iter->second;
+    }
   }
 
+  // For non-master zonegroups, enforce location constraint match
+  const bool enforce_location_match = !period || !my_zonegroup.is_master;
+  if (enforce_location_match && !my_zonegroup.equals(bucket_zonegroup->get_id())) {
```

Comment:
is there a reason that other zonegroups shouldn't also support creation of remote buckets?

in aws, it looks like you can send a [CreateBucket](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html) request to any region, naming any other valid region in the LocationConstraint

Reply:
No you will get `IllegalLocationConstraintException` is response. and tbh I would still be voting for master to also not create bucket for other zonegroups when the request is initiated by the user - not forwarded.
This would make some races easier to solve deleting the condition that we need to cover.

Reply:
> +1

:face_holding_back_tears: 


I just drafted a test case for this as well, will try to verify on my local asap.

Reply:
The tests have passed on my local, but we will need https://github.com/ceph/ceph/pull/59960.

Reply:
> but we will need #59960

working on it

Reply:
> > but we will need #59960
> 
> working on it

If you are running QA on that, perhaps it might be worth including this as well and dropping the test case there, as this case here is more accurate and covers both (creating a bucket and ensuring the bucket has been created for the corrected zonegroup).

Reply:
i cherry-picked the commit from this PR on top of the qa branch for https://github.com/ceph/ceph/pull/59960 and removed its test case. https://pulpito.ceph.com/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/ shows a similar failure for the two-zonegroup job (the two-zone job hasn't run yet):
```
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:======================================================================
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:ERROR: rgw_multi.tests.test_bucket_create_location_constraint
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2024-12-19T18:26:36.927 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_21fbe25bb2cf151e2eb892c1d86880b6eb21b2b5/qa/../src/test/rgw/rgw_multi/tests.py", line 3651, in test_bucket_create_location_constraint
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    z.s3_client.create_bucket(Bucket=bucket_name, CreateBucketConfiguration={'LocationConstraint': zg.name})
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/botocore/client.py", line 553, in _api_call
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    return self._make_api_call(operation_name, kwargs)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_teuthology_e4eeebda7d988dc20ee63a4f7e7d82e07061f6d5/virtualenv/lib/python3.10/site-packages/botocore/client.py", line 1009, in _make_api_call
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:    raise error_class(parsed_response, operation_name)
2024-12-19T18:26:36.928 INFO:tasks.rgw_multisite_tests:botocore.exceptions.ClientError: An error occurred (InvalidLocationConstraint) when calling the CreateBucket operation: The specified location-constraint is not valid
```

Reply:
By any chance, do we have access to rgw logs?

Reply:
zone a1: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi043/log/rgw.c1.client.0.log.gz
zone a2: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi104/log/rgw.c1.client.1.log.gz
zone b1: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi160/log/rgw.c2.client.0.log.gz
zone b2: https://qa-proxy.ceph.com/teuthology/cbodley-2024-12-19_17:37:17-rgw:multisite-wip-69281-distro-default-smithi/8045098/remote/smithi187/log/rgw.c2.client.1.log.gz

Reply:
I'm thinking maybe there is something fishy about the `zonegorups_by_api` var.
Currently, there are only two places the variable is getting filled, `RGWPeriodMap::update()` and `RGWPeriodMap::decode()`. I guess decode is irrelevant here so we should see where we have `RGWPeriodMap::update()` called.
I guess it should be `RGWPeriod::update()` but I can't find any logs saying `RGWPeriod::update() realm ...` due to the second line of the function logging with no condition.
the other possibility is `update_period()` which is only called by rgw-admin where perhaps on `period update --commit` commands it should construct that for itself but still on the master zonegorup and master zone `a1` (which we are sure the zonegroup `a` has been created) it cannot find itself in the period map
```
2024-12-19T18:26:36.873+0000 7f1be8535640  0 req 13354008642218923451 0.000000000s s3:create_bucket location constraint (a) can't be found.
```

Reply:
ok this is actually weird - api_name is empty.
```
2024-12-19T18:15:53.211 INFO:teuthology.orchestra.run.smithi043.stdout:{"id":"8987fccd-6980-4682-9cce-cf39a4aa6b34","epoch":12,"predecessor_uuid":"5276a8ed-77a4-4898-9669-84e295758977","sync_status":[],"period_map":{"id":"8987fccd-6980-4682-9cce-cf39a4aa6b34","zonegroups":[{"id":"a","name":"a","api_name":"","is_master":true,"endpoints":["http://smithi043.front.sepia.ceph.com:8000"],"hostnames":[],"hostnames_s3website":[],"master_zone":"6a93f239-7b93-4dfe-94b1-baf12183eb58","zones":[{"id":"250dd5e0-4049-46ca-bf5b-e7e54b3b29b4","name":"a2","endpoints":["http://smithi104.front.sepia.ceph.com:8001"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]},{"id":"6a93f239-7b93-4dfe-94b1-baf12183eb58","name":"a1","endpoints":["http://smithi043.front.sepia.ceph.com:8000"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]}],"placement_targets":[{"name":"default-placement","tags":[],"storage_classes":["STANDARD"]}],"default_placement":"default-placement","realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","sync_policy":{"groups":[]},"enabled_features":["notification_v2","resharding"]},{"id":"b","name":"b","api_name":"","is_master":false,"endpoints":["http://smithi160.front.sepia.ceph.com:8000"],"hostnames":[],"hostnames_s3website":[],"master_zone":"b3be8bac-85cd-4275-8c15-00f60f318a67","zones":[{"id":"b3be8bac-85cd-4275-8c15-00f60f318a67","name":"b1","endpoints":["http://smithi160.front.sepia.ceph.com:8000"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]},{"id":"e868f7a7-a7b3-4202-8e6f-86642035ca92","name":"b2","endpoints":["http://smithi187.front.sepia.ceph.com:8001"],"log_meta":false,"log_data":true,"bucket_index_max_shards":11,"read_only":false,"tier_type":"","sync_from_all":true,"sync_from":[],"redirect_zone":"","supported_features":["compress-encrypted","notification_v2","resharding"]}],"placement_targets":[{"name":"default-placement","tags":[],"storage_classes":["STANDARD"]}],"default_placement":"default-placement","realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","sync_policy":{"groups":[]},"enabled_features":["notification_v2","resharding"]}],"short_zone_ids":[{"key":"250dd5e0-4049-46ca-bf5b-e7e54b3b29b4","val":104308217},{"key":"6a93f239-7b93-4dfe-94b1-baf12183eb58","val":2854171449},{"key":"b3be8bac-85cd-4275-8c15-00f60f318a67","val":1981456006},{"key":"e868f7a7-a7b3-4202-8e6f-86642035ca92","val":3492425437}]},"master_zonegroup":"a","master_zone":"6a93f239-7b93-4dfe-94b1-baf12183eb58","period_config":{"bucket_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1},"user_quota":{"enabled":false,"check_on_raw":false,"max_size":-1,"max_size_kb":0,"max_objects":-1},"user_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false},"bucket_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false},"anonymous_ratelimit":{"max_read_ops":0,"max_write_ops":0,"max_read_bytes":0,"max_write_bytes":0,"enabled":false}},"realm_id":"a54451a6-0f7f-48fa-b6c0-a34c0eb1ea9d","realm_epoch":2}
```

Reply:
hmm.. I'm not sure why we create zonegroups via `zonegroup set` but that might explain why api_name is created empty https://github.com/ceph/ceph/blob/862ed6e9cd8cece7edc6d7d45505b5a4b4f00e26/qa/tasks/rgw_multisite.py#L357-L369

Reply:
@cbodley I pushed a fix (c21b5f7e0b1a48e7a4b4e21d1d357e8f2f26a1ac), can you try again?